plugins/uartdmx/UartDmxThread.cpp

```diff
@@ -93,46 +105,94 @@ void *UartDmxThread::Run() {
       buffer.Set(m_buffer);
     }
 
-    if (!m_widget->SetBreak(true))
-      goto framesleep;
+    writeDmxData(buffer);
+  }
+  return NULL;
+}
 
-    if (m_granularity == GOOD)
-      usleep(m_breakt);
+/**
+ * Write the DMX data to the actual UART interface.
+ */
+void UartDmxThread::writeDmxData(const DmxBuffer &buffer) {
+  TimeStamp ts1;
+  Clock clock;
 
-    if (!m_widget->SetBreak(false))
-      goto framesleep;
+  // ensures that ts1 has a valid value for the framesleep if break condition cannot be locked/released
+  clock.CurrentMonotonicTime(&ts1);
 
-    if (m_granularity == GOOD)
-      usleep(DMX_MAB);
+  if (!m_widget->SetBreak(true)) {
+    frameSleep(ts1);
+    return;
+  }
 
-    if (!m_widget->Write(buffer))
-      goto framesleep;
+  if (m_granularity == GOOD)
+    usleep(m_breakt);
 
-  framesleep:
-    // Sleep for the remainder of the DMX frame time
-    usleep(m_malft);
+  if (!m_widget->SetBreak(false)) {
+    frameSleep(ts1);
+    return;
   }
-  return NULL;
+
+  // stores the time stamp for the elapsed time calculation during the framesleep
+  clock.CurrentMonotonicTime(&ts1);
+
+  if (m_granularity == GOOD)
+    usleep(DMX_MAB);
+
+  m_widget->Write(buffer);
+  
+  frameSleep(ts1);
 }
 
+/**
+ * Sleeps for the rest of the frame time and tries to recover the granularity if needed. 
+ */
+void UartDmxThread::frameSleep(const TimeStamp &ts1) {
+  TimeStamp ts2, ts3;
+  Clock clock;
+
+  // Sleep for the remainder of the DMX frame time
+  clock.CurrentMonotonicTime(&ts2);
+  TimeInterval elapsed = ts2 - ts1;
+
+  if (m_granularity == GOOD) {
+    while (elapsed.InMilliSeconds() < m_frameTime) {
+      usleep(1000);
```

Comment:
Similar comment, and I suspect this is previous code, but is it a magic value or just a random second, I think it's the max break to break to ensure we start afresh, but...

Reply:
I replaced that with a constant and added a comment for better understanding